main/eval/src/mill/eval/Evaluator.scala

```diff
@@ -750,225 +45,64 @@ object Evaluator {
-                  t.ctx.enclosing.split("[.# ]").map(Segment.Label)
-              )
-            }
-          )
-        )
-      case t if goals.contains(t) => Left(t)
+    case class Task[T](task: mill.define.Task[_]) extends Terminal {
+      def render = task.toString
     }
-
-    (sortedGroups, transitive)
   }
 
-  case class Evaluated(
-      newResults: collection.Map[Task[_], TaskResult[(Val, Int)]],
-      newEvaluated: Seq[Task[_]],
-      cached: Boolean
-  ) {
-    private[Evaluator] def copy(
-        newResults: collection.Map[Task[_], TaskResult[(Val, Int)]],
-        newEvaluated: Seq[Task[_]] = newEvaluated,
-        cached: Boolean = cached
-    ): Evaluated = new Evaluated(newResults, newEvaluated, cached)
-  }
-
-  object Evaluated {
-    private[Evaluator] def unapply(evaluated: Evaluated)
-        : Option[(collection.Map[Task[_], TaskResult[(Val, Int)]], Seq[Task[_]], Boolean)] =
-      Some((evaluated.newResults, evaluated.newEvaluated, evaluated.cached))
-  }
-
-  // Increment the counter message by 1 to go from 1/10 to 10/10
-  class NextCounterMsg(taskCount: Int) {
-    var counter: Int = 0
+  trait Results {
+    def rawValues: Seq[Result[Val]]
+    def evaluated: Agg[Task[_]]
+    def transitive: Agg[Task[_]]
+    def failing: MultiBiMap[Terminal, Result.Failing[Val]]
+    def results: collection.Map[Task[_], TaskResult[Val]]
 
-    def apply(): String = {
-      counter += 1
-      s"${counter}/${taskCount}"
-    }
+    def values: Seq[Val] = rawValues.collect { case Result.Success(v) => v }
   }
 
-  def writeTracings(tracings: Seq[TraceEvent], outPath: os.Path): Unit = {
-    os.write.over(
-      outPath / "mill-par-profile.json",
-      upickle.default.stream(tracings, indent = 2)
+  case class TaskResult[T](result: Result[T], recalc: () => Result[T]) {
+    def map[V](f: T => V) = TaskResult[V](
+      result.map(f),
+      () => recalc().map(f)
     )
   }
 
-  class EvalOrThrow(evaluator: Evaluator, exceptionFactory: Results => Throwable) {
-    def apply[T: ClassTag](task: Task[T]): T =
-      evaluator.evaluate(Agg(task)) match {
-        case r if r.failing.items().nonEmpty =>
-          throw exceptionFactory(r)
-        case r =>
-          // Input is a single-item Agg, so we also expect a single-item result
-          val Seq(Val(e: T)) = r.values
-          e
-      }
-    def apply[T: ClassTag](tasks: Seq[Task[T]]): Seq[T] =
-      evaluator.evaluate(tasks) match {
-        case r if r.failing.items().nonEmpty =>
-          throw exceptionFactory(r)
-        case r => r.values.map(_.value).asInstanceOf[Seq[T]]
-      }
-  }
+  // This needs to be a ThreadLocal because we need to pass it into the body of
+  // the TargetScopt#read call, which does not accept additional parameters.
+  // Until we migrate our CLI parsing off of Scopt (so we can pass the BaseModule
+  // in directly) we are forced to pass it in via a ThreadLocal
+  val currentEvaluator = new ThreadLocal[mill.eval.Evaluator]
```

Comment:
How about making this a `scala.util.DynamicVariable`, which uses a `ThreadLocal` internally, but handles re-setting and cleanup more safely through the use of `withValue`?

Reply:
done